File: android/AVL/AVLTree.java
package com.example.rolando.myapplication_buenaversiob.AVL;

/**
 * Created by Rolando on 10/11/2016.
 */
public class AVLTree <E> {
    AVLNode root = null;
    
    Integer count = 0;
    public boolean recursive_insert(AVLNode subroot,E data){
        if (((Comparable)data).compareTo(subroot.data)==0){
            return false;
        }
        if (((Comparable)data).compareTo(subroot.data)<0){
            if (subroot.left == null){
                subroot.left = new AVLNode<E>(subroot,data);
                count++;
                rebalance(subroot);
                return true;
            }else{
                return recursive_insert(subroot.left,data);
            }
        }else{
            if (subroot.right == null){
                subroot.right = new AVLNode<E>(subroot,data);
                count++;
                rebalance(subroot);
                return  true;
            }else{
                return recursive_insert(subroot.right,data);
            }
        }

    }
    private boolean recursive_delete(AVLNode node, AVLNode dummy_root, E data) {
        if (((Comparable)data).compareTo(node.data)<0){
            if (node.left != null){
                recursive_delete(node.left,node,data);
            }else{
                return false;
            }
        }else if (((Comparable)data).compareTo(node.data)>0){
            if (node.right != null){
                recursive_delete(node.right,node,data);
            }else{
                return false;
            }
        }else{
            if (node.left != null && node.right != null){
                node.setData(find_mi(node.right));
                recursive_delete(node.right,node,(E)node.data);
            }else if (dummy_root.left == node){
                dummy_root.left = (node.left != null)? node.left : node.right;
            }else if (dummy_root.right == node){
                dummy_root.right = (node.left != null)? node.left : node.right;
            }

        }
        rebalance(dummy_root);
        return true;
    }
    public E find_mi(AVLNode node){
        if(node != null){
            while (node.left != null){
                node = node.left;
            }
        }
        return (E) node.getData();
    }

    public void rebalance(AVLNode subroot){
        set_balance(subroot,subroot.left,subroot.right);
        if (subroot.balance == -2){
            if (subroot.left.balance <= 0){
                subroot = rotate_left_left(subroot);
            } else if (subroot.balance == 1){
                subroot = rotate_left_right(subroot);
            }
        } else if (subroot.balance == 2){
            if (subroot.right.balance >= 0){
                subroot = rotate_right_right(subroot);
            }else if (subroot.balance == -1){
                subroot = rotate_right_left(subroot);
            }
        }
        if (subroot.parent != null){
            rebalance(subroot.parent);
        }else{
            this.root = subroot;
        }
    }

    public AVLNode rotate_right_left(AVLNode subroot){
        subroot.left = rotate_left_left(subroot.left);
        return rotate_right_right(subroot);
    }

    public AVLNode rotate_left_right(AVLNode subroot){
        subroot.right = rotate_right_right(subroot.right);
        return rotate_left_left(subroot);
    }

    public AVLNode rotate_left_left(AVLNode subroot){
        AVLNode aux = subroot.left;
        aux.parent = subroot.parent;
        subroot.left = aux.right;
        if (subroot.left != null){
            subroot.left.parent = subroot;
        }
        aux.right = subroot;
        subroot.parent = aux;
        if (aux.parent != null){
            if (aux.parent.right == subroot){
                aux.parent.right = aux;
            }else{
                aux.parent.left = aux;
            }
        }
        set_balance(subroot, aux);
        return aux;
    }

    public AVLNode rotate_right_right(AVLNode subroot){
        AVLNode aux = subroot.right;
        aux.parent = subroot.parent;
        subroot.right = aux.left;
        if (subroot.right != null){
            subroot.right.parent = subroot;
        }
        aux.left = subroot;
        subroot.parent = aux;
        if (aux.parent != null){
            if (aux.parent.right == subroot){
                aux.parent.right = aux;
            }else{
                aux.parent.left = aux;
            }
        }
        set_balance(subroot, aux);
        return aux;
    }

    public void set_balance(AVLNode... nodes){
        for(AVLNode node : nodes){
            if (node != null)
                node.balance = height(node.right) - height(node.left);
        }
    }
    public Integer height(AVLNode subroot){
        if(subroot == null){
            return -1;
        }else{
            return 1 + Math.max(height(subroot.left),height(subroot.right));
        }
    }

    public boolean insert(E data) {
        if (root == null){
            root = new AVLNode(data);
            root.setData(data);
        }else{
            recursive_insert(root,data);
        }
        return false;
    }

    public void inorder(AVLNode node) {
        if(node!=null){
            inorder(node.getLeft());
            System.out.println(node.getData());
            inorder(node.getRight());
        }
    }


}

File: web/data-structures/AVLTree.php
<?php 
require_once 'AVLNode.php';

class AVLTree {
  public $root;

  /**
   * 
   * Insert
   * 
   * Inserts a data into the Binary Search Tree.
   * Returns nothing.
   * 
   */
  function insert($data) {
    if ($this->root == null) {
      $this->root = new AVLNode(null, $data, null, null);
    } else {
      $this->recursiveInsert($this->root, $data);
    }
  }

  private function recursiveInsert($subroot, $data) {
    if ($data == $subroot->data) {
      return false;
    } elseif ($data < $subroot->data) {

      if ($subroot->left == null) {
        $subroot->left = new AVLNode($subroot, $data, null, null);
        // count++;
        #$this->rebalance($subroot);

        return true;
      } else {
        $this->recursiveInsert($subroot->left, $data);
      }

    } else {

      if ($subroot->right == null) {
        $subroot->right = new AVLNode($subroot, $data, null, null);
        // count++;
        #$this->rebalance($subroot);
        
        return true;
      } else {
        $this->recursiveInsert($subroot->right, $data);
      }

    }
  }

  /**
   * 
   * Delete
   * 
   * Delete a data in the tree.
   * Returns a boolean.
   * 
   */
  function delete($data) {
    if ($this->root == null) {
      return false;
    } else {

      if ($data == $this->root->data) {
        $dummyRoot = AVLNode::withData(null);
        $dummyRoot->left = $this->root;
        $result = $this->recursiveDelete($this->root, $dummyRoot, $data);
        $this->root = $dummyRoot->left;
        return $result;
      } else {
        return $this->recursiveDelete($this->root, null, $data);
      }

    }

  }

  function recursiveDelete($subroot, $parent, $data) {
    if ($data < $subroot->data) {
      
      if ($subroot->left != null) {
        $this->recursiveDelete($subroot->left, $subroot, $data);
      } else {
        return false;
      }

    } elseif ($data > $subroot->data) {
      
      if ($subroot->right != null) {
        $this->recursiveDelete($subroot->right, $subroot, $data);
      } else {
        return false;
      }

    } else {

      if ($subroot->left != null && $subroot->right != null) {
        $subroot->data = $this->findMin($subroot->right);
        $this->recursiveDelete($subroot->right, $subroot, $data);
      } elseif ($parent->left == $subroot) {
        $parent->left = ($subroot->left != null) ? $subroot->left : $subroot->right;
      } elseif ($parent->right == $subroot) {
        $parent->right = ($subroot->left != null) ? $subroot->left : $subroot->right;
      }
      #$this->rebalance($parent);

      return true;
    }

  }

  /**
   * 
   * Contains
   * 
   * Check if the tree contains a specified data.
   * Returns a boolean.
   * 
   */
  function contains($data) {
    return $this->recursiveContains($this->root, $data);
  }

  private function recursiveContains($subroot, $data) {
    if ($subroot != null) {
      
      if ($data == $subroot->data) {
        return true;
      } else {

        if ($data < $subroot->data) {
          return $this->recursiveContains($subroot->left, $data);
        } else {
          return $this->recursiveContains($subroot->right, $data);
        }

      }

    } else {
      return false;
    }
  }

  /**
   * 
   * Rebalance
   * 
   * Check the balance of the tree and make a rebalance if 
   * it's necesary.
   * Returns nothing.
   * 
   */
  private function rebalance($subroot) {
    $this->setBalance($subroot, $subroot->left, $subroot->right);
    //echo "<pre>".print_r($subroot)."</pre>";

    if ($subroot->balance == -2) {
      
      if ($subroot->left->balance <= 0) {
        $subroot = $this->rotateLeftLeft($subroot);
      } elseif ($subroot->balance == 1) {
        $subroot = $this->rotateLeftRight($subroot);
      }

    } elseif ($subroot->balance == 2) {

      if ($subroot->right->balance >= 0) {
        $subroot = $this->rotateRightRight($subroot);
      } elseif ($subroot->balance == -1) {
        $subroot = $this->rotateRightLeft($subroot);
      }

    }

    if ($subroot->parent != null) {
      $this->rebalance($subroot->parent);
    } else {
      $this->root = $subroot; 
    }
  }

  private function setBalance( ...$nodes) {
    foreach ($nodes as $node) {
      if ($node != null) {
        $node->balance = ( $this->height($node->right) ) - ( $this->height($node->left) );
      }
    }
  }
  
  private function rotateLeftLeft($subroot) {
    //var_dump($subroot);
    $aux = $subroot->left;
    $aux->parent = $subroot->parent;
    $subroot->left = $aux->right;
    
    if ($subroot->left == null) {
      $subroot->left->parent = $subroot;
    }
    
    $aux->right = $subroot;
    $subroot->parent = $aux;

    if ($aux->parent != null) {
      if ($aux->parent->right == $subroot) {
        $aux->parent->right = $aux;
      } else {
        $aux->parent->left = $aux;
      }
    }

    $this->setBalance($subroot, $aux);
    return $aux;
  }

  private function rotateRightRight($subroot) {
    $aux = $subroot->right;
    $aux->parent = $subroot->parent;
    $subroot->right = $aux->left;
    
    if ($subroot->right != null) {
      $subroot->right->parent = $subroot;
    }
    
    $aux->left = $subroot;
    $subroot->parent = $aux;

    if ($aux->parent != null) {
      if ($aux->parent->right == $subroot) {
        $aux->parent->right = $aux;
      } else {
        $aux->parent->left = $aux;
      }
    }

    $this->setBalance($subroot, $aux);
    return $aux;
  }

  private function rotateRightLeft($subroot) {
    $subroot->left = $this->rotateLeftLeft($subroot->left);
    return $this->rotateRightRight($subroot);
  }

  private function rotateLeftRight($subroot) {
    $subroot->right = $this->rotateRightRight($subroot->right);
    return $this->rotateLeftLeft($subroot);
  }

  function height($subroot) {
    if ($subroot == null) {
      return -1;
    } else {
      return 1 + max($this->height($subroot->left), $this->height($subroot->right));
    }
  }

  /**
   * 
   * More helpers
   * 
   */
  private function findMin($subroot) {
    if ($subroot != null) {
      while ($subroot->left != null) {
        $subroot = $subroot->left;
      }
    }
    return $subroot->data;
  }
  
  private function findMax($subroot) {
    if ($subroot != null) {
      while ($subroot->right != null) {
        $subroot = $subroot->right;
      }
    }
    return $subroot->data;
  }

  /**
   * 
   * Tree transversals:
   * Preorder, inorder and postorder
   * 
   */
  function preorder() {
    $array = array();
    $array = $this->recursivePreorder($array, $this->root);
    return $array;
  }

  private function recursivePreorder($array, $node) {
    if ($node != null) {
      array_push($array, $node->data);
      $array = $this->recursivePreorder($array, $node->left);
      $array = $this->recursivePreorder($array, $node->right);
    }
    return $array;
  }

  function inorder() {
    $array = array();
    $array = $this->recursiveInorder($array, $this->root);
    return $array;
  }

  private function recursiveInorder($array, $node) {
    if ($node != null) {
      $array = $this->recursiveInorder($array, $node->left);
      array_push($array, $node->data);
      $array = $this->recursiveInorder($array, $node->right);
    }
    return $array;
  }

  function postorder() {
    $array = array();
    $array = $this->recursivePostorder($array, $this->root);
    return $array;
  }

  private function recursivePostorder($array, $node) {
    if ($node != null) {
      $array = $this->recursivePostorder($array, $node->left);
      $array = $this->recursivePostorder($array, $node->right);
      array_push($array, $node->data);
    }
    return $array;
  }



}


 ?>